### PR TITLE
Xilinx: Update examples

### PR DESCRIPTION
**Pull request by edmooring**
Update the load_fw demo for Xilinx to use new power management APIs.
Update other demos to incorporate new Xilinx BSP.
Most of this code is specific to the Xilinx hardware, but there are a few fixes to generic code.

Signed-off-by: Ed Mooring <ed.mooring@linaro.org>

### CODE REVIEW

File: apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c [hunk 1]
```diff
@@ -9,13 +9,13 @@
 
 #include <common.h>
 
+
```
Comment (arnopo) [inline apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c]: clean extra line

File: apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c [hunk 2]
```diff
@@ -152,8 +159,14 @@ static int apu_rproc_shutdown(struct remoteproc *rproc)
 		/* Release TCM resource */
 		pa = mem->pa;
 		pa_end = metal_io_phys(mem->io, metal_io_region_size(mem->io));
-		if (pa_end <= 0x7FFFFFFF)
+		if (pa_end <= 0x7FFFFFFF){
```
Comment (arnopo) [inline apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c]: typo: space after )

File: apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c [hunk 3]
```diff
@@ -9,13 +9,13 @@
...
 
+
 static struct remoteproc *apu_rproc_init(struct remoteproc *rproc,
 				 struct remoteproc_ops *ops, void *arg)
 {
 	struct rproc_priv *priv;
 	unsigned int cpu_id = *((unsigned int *)arg);
-
-	if (cpu_id < NODE_APU_0 || cpu_id > NODE_APU_3){
+	if(cpu_id < APU_NODE_0 || cpu_id > APU_NODE_1){
```
Comment (arnopo) [inline apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c]: typo: space after )

File: apps/machine/zynqmp/platform_info.c [hunk 4]
```diff
@@ -34,15 +34,20 @@
 #define RPU_CPU_ID          0 /* RPU remote CPU Index. We only talk to
                                * one CPU in the exmaple. We set the CPU
                                * index to 0. */
-#define IPI_CHN_BITMASK     0x00000100 /* IPI channel bit mask for IPI
+#ifdef versal
+#define IPI_CHN_BITMASK     0x08 /* IPI channel bit mask for IPI
```
Comment (arnopo) [inline apps/machine/zynqmp/platform_info.c]: realign comment above

File: apps/machine/zynqmp/platform_info.c [hunk 5]
```diff
@@ -34,15 +34,20 @@
...
 					* from/to RPU0 */
+#define IPI_DEV_NAME        "ff360000.ipi" /* IPI device name */
+#else
+#define IPI_CHN_BITMASK		0x00000100
+#define IPI_DEV_NAME		"ff340000.ipi"
+#endif /* versal */
```
Comment (arnopo) [inline apps/machine/zynqmp/platform_info.c]: nitpicking: define could be aligned

File: apps/machine/zynqmp_r5/platform_info.h [hunk 6]
```diff
@@ -17,7 +17,16 @@ extern "C" {
...
+#ifdef versal
+#define IPI_IRQ_VECT_ID		63
+#define IPI_BASE_ADDR       0xFF340000 /* IPI base address*/
+#define IPI_CHN_BITMASK     0x0000020 /* IPI channel bit mask for IPI from/to
+					   APU */
+#else
 #define IPI_IRQ_VECT_ID         XPAR_XIPIPSU_0_INT_ID
+#define IPI_BASE_ADDR		XPAR_XIPIPSU_0_BASE_ADDRESS
+#define IPI_CHN_BITMASK		0x01000000
+#endif /* versal */
```
Comment (arnopo) [inline apps/machine/zynqmp_r5/platform_info.h]: alignment for readability

File: apps/system/generic/machine/zynqmp_r5/CMakeLists.txt [hunk 7]
```diff
@@ -12,6 +12,21 @@ find_library(LIBXIL_LIB NAMES xil PATHS ${CMAKE_FIND_ROOT_PATH})
...
+if (NOT VERSION_1_PM_CLIENT)
+  # check for version 2
+  find_path(VERSION_2_PM_CLIENT "xillibpm_api.h")
+  if (NOT VERSION_2_PM_CLIENT)
+    message(FATAL_ERROR "PM Client not found")
+  else()
+   collect (APP_EXTRA_C_FLAGS "  -DVERSION_2_PM_CLIENT ")
+  endif()
+else()
+ collect (APP_EXTRA_C_FLAGS "  -DVERSION_1_PM_CLIENT ")
```
Comment (arnopo) [inline apps/system/generic/machine/zynqmp_r5/CMakeLists.txt]: indentation

File: apps/system/generic/machine/zynqmp_r5/CMakeLists.txt [hunk 8]
```diff
@@ -12,6 +12,21 @@ find_library(LIBXIL_LIB NAMES xil PATHS ${CMAKE_FIND_ROOT_PATH})
...
+# check PM API for certain headers and set client version based
+# on which functions are found
+find_path(VERSION_1_PM_CLIENT "pm_api_sys.h")
+if (NOT VERSION_1_PM_CLIENT)
+  # check for version 2
+  find_path(VERSION_2_PM_CLIENT "xillibpm_api.h")
+  if (NOT VERSION_2_PM_CLIENT)
+    message(FATAL_ERROR "PM Client not found")
+  else()
+   collect (APP_EXTRA_C_FLAGS "  -DVERSION_2_PM_CLIENT ")
```
Comment (arnopo) [inline apps/system/generic/machine/zynqmp_r5/CMakeLists.txt]: indentation

File: apps/system/linux/machine/generic/platform_info.c [hunk 9]
```diff
@@ -270,6 +283,12 @@ linux_proc_init(struct remoteproc *rproc,
 	return rproc;
 
 err:
+	if (sg != NULL) {
+		metal_shmem_munmap(metal_shm, sg);
+	}
+	if (metal_shm != NULL) {
+		metal_shmem_close(metal_shm);
+	}
```
[the remark this answers is not in the document]
Comment (edmooring) [reply apps/system/linux/machine/generic/platform_info.c]: Good question. The function is already returning NULL for the error case. If a lazy programmer failed to check the error return, then setting those elements to NULL might make the program fail a bit "louder".

I don't think it's worth it in this case. The extra assignments might make the intent of the code less clear.

File: apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+{
+	struct rproc_priv *priv;
+	unsigned int cpu_id = *((unsigned int *)arg);
+
+	if (cpu_id < NODE_APU_0 || cpu_id > NODE_APU_3){
+		xil_printf("%s: invalide node id: %d \n\r",__func__, cpu_id);
+		return NULL;
+	}
+
+	xil_printf("%s: node id: %d\n\r", __func__, cpu_id);
```
[the remark this answers is not in the document]
Comment (edmooring) [reply apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c]: All of this code uses xil_printf() under the covers. Changed these instances to use LPERROR() and LPRINTF() for stylistic consistency.

File: apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+		return NULL;
+
+	LPRINTF("%s: pa=0x%x, da=0x%x, size=0x%x, atrribute=0x%x\n\r",
+		__func__, *pa, *da, size, attribute);
+	lda = *da;
+	lpa = *pa;
+	if (!attribute)
+		attribute = NORM_SHARED_NCACHE | PRIV_RW_USER_RW;
+
+	metal_phys_addr_t lda_end;
```
[the remark this answers is not in the document]
Comment (edmooring) [reply apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c]: Agreed.

File: apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+			REQUEST_ACK_BLOCKING);
+	if (lda_end >= 0xFFFF0000)
+		XPm_RequestNode(NODE_OCM_BANK_3,
+			PM_CAP_ACCESS, 0,
+			REQUEST_ACK_BLOCKING);
+
+	if (lpa == METAL_BAD_PHYS)
+		lpa = lda;
+	if (lpa == METAL_BAD_PHYS)
+		return NULL;
```
[the remark this answers is not in the document]
Comment (edmooring) [reply apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c]: Done.

File: apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+	priv->rproc->ops = ops;
+	metal_list_init(&priv->rproc->mems);
+	priv->rproc->priv = priv;
+	rproc->state = RPROC_READY;
+	return priv->rproc;
+}
+
+static void apu_rproc_remove(struct remoteproc *rproc)
+{
+	if (rproc) {
```
[the remark this answers is not in the document]
Comment (edmooring) [reply apps/examples/load_fw/zynqmp_apu_lcm_rproc_example.c]: It's actually already checked in the one place that calls it (remoteproc_remove), so this check is redundant.

**Comment by edmooring** [comment]
> @edmooring
> Could you clean the "merge commit" in normal commit? This would simplify the review process...
> some minor remarks
> Thanks!
The merge commits just add a little "noise" to the log. The actual commits, with their original IDs, are preserved in the history.

Merge 39deda merged a0b2ce and 3f6df5 merged 40a94a, so you can just skip the merge commits.

**Comment by arnopo** [comment]
Hi Ed,
For the review process could you  update directly the patches according to the reviews comment, instead of sending fixes on top? To send a V2 you have just to modify your patches and force your github branch on this new version. Then your OpenAMP pull request will be automatically updated,  and comments treated will be tagged as resolved (or outdated). 
This would simplify the review.
Thanks,
Arnaud

**Pull request merged**